src/libraries/System.Net.Sockets/tests/FunctionalTests/DisposedSocketTests.cs

```diff
@@ -761,6 +761,16 @@ public async Task NonDisposedSocket_SafeHandlesCollected(bool clientAsync)
             });
         }
 
+        [Fact]
+        public void SocketWithDanglingReferenceDoesntHangFinalizerThread()
+        {
+            Socket socket = new Socket(SocketType.Stream, ProtocolType.Tcp);
+            bool dummy = false;
+            socket.SafeHandle.DangerousAddRef(ref dummy);
+            GC.Collect();
```

Comment:
This will not collect `socket` since it's still being referenced by the current stack frame. The creation has to be moved to a non-inlined method.

Not 100% sure about this idea of mine, but I would even consider merging this test with `NonDisposedSocket_SafeHandlesCollected` to stress the case more against future regressions:

```C#
[ConditionalTheory(typeof(PlatformDetection), nameof(PlatformDetection.IsPreciseGcSupported))]
[InlineData(false, false)]
[InlineData(true, false)]
[InlineData(false, true)]
[InlineData(true, true)]
public async Task NonDisposedSocket_SafeHandlesCollected(bool clientAsync, bool dangerousAddRef)
{
    List<WeakReference> handles = await CreateHandlesAsync(clientAsync, dangerousAddRef);
    RetryHelper.Execute(() =>
    {
        GC.Collect();
        GC.WaitForPendingFinalizers();
        Assert.Equal(0, handles.Count(h => h.IsAlive));
    });
}

[MethodImpl(MethodImplOptions.NoInlining)]
private static async Task<List<WeakReference>> CreateHandlesAsync(bool clientAsync, bool dangerousAddRef)
{
    var handles = new List<WeakReference>();

    using (var listener = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp))
    {
        listener.Bind(new IPEndPoint(IPAddress.Loopback, 0));
        listener.Listen();

        for (int i = 0; i < 100; i++)
        {
            var client = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp); // do not dispose

            if (dangerousAddRef)
            {
                bool addRefSuccess = false;
                client.SafeHandle.DangerousAddRef(ref addRefSuccess);
                Assert.True(addRefSuccess);
            }

            handles.Add(new WeakReference(client.SafeHandle));
            if (clientAsync)
            {
                await client.ConnectAsync(listener.LocalEndPoint);
            }
            else
            {
                client.Connect(listener.LocalEndPoint);
            }

            using (Socket server = listener.Accept())
            {
                if (clientAsync)
                {
                    Task<int> receiveTask = client.ReceiveAsync(new ArraySegment<byte>(new byte[1]), SocketFlags.None);
                    Assert.Equal(1, server.Send(new byte[1]));
                    Assert.Equal(1, await receiveTask);
                }
                else
                {
                    Assert.Equal(1, server.Send(new byte[1]));
                    Assert.Equal(1, client.Receive(new byte[1]));
                }
            }
        }
    }

    return handles;
}
```

Reply:
> This will not collect socket since it's still being referenced by the current stack frame. The creation has to be moved to a non-inlined method.

Good catch!

> Not 100% sure about this idea of mine, but I would even consider merging this test with NonDisposedSocket_SafeHandlesCollected to stress the case more against future regressions:

I kept them separate atm because I find they are testing distinct cases (APIs used, and covered code).